opal/mca/smsc/xpmem/smsc_xpmem_module.c

```diff
@@ -46,131 +48,160 @@ mca_smsc_endpoint_t *mca_smsc_xpmem_get_endpoint(opal_proc_t *peer_proc)
         return NULL;
     }
 
-    endpoint->apid = xpmem_get(modex->seg_id, XPMEM_RDWR, XPMEM_PERMIT_MODE, (void *) 0666);
-    endpoint->address_max = modex->address_max;
-
-    return &endpoint->super;
-}
-
-struct mca_smsc_xpmem_cleanup_reg_ctx_t {
-    mca_smsc_xpmem_endpoint_t *endpoint;
-    opal_list_t *registrations;
-};
-
-typedef struct mca_smsc_xpmem_cleanup_reg_ctx_t mca_smsc_xpmem_cleanup_reg_ctx_t;
-
-struct mca_smsc_xpmem_check_reg_ctx_t {
-    mca_smsc_xpmem_endpoint_t *endpoint;
-    mca_rcache_base_registration_t **reg;
-    uintptr_t base;
-    uintptr_t bound;
-};
-typedef struct mca_smsc_xpmem_check_reg_ctx_t mca_smsc_xpmem_check_reg_ctx_t;
-
-static int mca_smsc_xpmem_check_reg(mca_rcache_base_registration_t *reg, void *ctx)
-{
-    mca_smsc_xpmem_check_reg_ctx_t *xpmem_ctx = (mca_smsc_xpmem_check_reg_ctx_t *) ctx;
-
-    if (reg->alloc_base != (void *) xpmem_ctx->endpoint) {
-        /* ignore this registration */
-        return OPAL_SUCCESS;
+    endpoint->vma_module = mca_rcache_base_vma_module_alloc();
+    if (OPAL_UNLIKELY(NULL == endpoint->vma_module)) {
+        OBJ_RELEASE(endpoint);
```

Comment:
I moved the modex operation earlier to avoid one such instance. But I'm not sure I can elegantly avoid the release(s) in the other 2 fail blocks, as all 3 operations (ep, vma_module, xpmem_get) may fail and all need cleanup.